### PR TITLE
Improvement/fluent builders

### PR DESCRIPTION
**Pull request by SgtSilvio**
**Motivation**
Make writing of MQTT message builders more fluent and simpler with nested builders while still remaining the opportunity to pass already built objects.

**Changes**
- Added FluentBuilder
- Extended FluentBuilder in builders for:
  - MQTT messages (3 & 5)
  - Topic, TopicFilter/SharedTopicFilter
  - UserProperties
  - Ssl/Websocket/ExecutorConfig
- Added fluent builder methods (simpleAuth, restrictions, willPublish, addSubscription, addTopicFilter, topic, topicFilter, userProperties, useSsl, useWebsocket, executorConfig)

### CODE REVIEW

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java [hunk 1]
```diff
@@ -22,29 +22,37 @@
...
 
 import static org.mqttbee.mqtt.message.publish.MqttWillPublish.DEFAULT_DELAY_INTERVAL;
 
 /**
  * @author Silvio Giebl
  */
-public class Mqtt5WillPublishBuilder extends Mqtt5PublishBuilder {
+public class Mqtt5WillPublishBuilder<P> extends Mqtt5PublishBuilder<P> {
+
+    private final Function<? super Mqtt5WillPublish, P> parentConsumer;
```
Comment (christianhoff) [inline src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java]: why is this needed?

File: src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java [hunk 2]
```diff
@@ -18,35 +18,53 @@
...
+
+import java.util.function.Function;
 
 /**
  * @author Silvio Giebl
  */
-public class MqttSharedTopicFilterBuilder extends MqttTopicFilterBuilder {
+public class MqttSharedTopicFilterBuilder<P> extends MqttTopicFilterBuilder<P> {
+
+    private final Function<? super MqttSharedTopicFilter, P> parentConsumer;
```
[the remark this answers is not in the document]
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java]: Because MqttTopicFilterBuilder overrides MqttSharedTopicFilterBuilder
so we need `Function<? super MqttSharedTopicFilter, P>`
but we inherit `Function<? super MqttTopicFilter, P>`

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java (hunk 1 above)
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java]: Because MqttWillPublishBuilder extends MqttPublishBuilder

File: src/main/java/org/mqttbee/util/FluentBuilder.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+package org.mqttbee.util;
+
+import org.mqttbee.annotations.NotNull;
+import org.mqttbee.annotations.Nullable;
+
+import java.util.function.Function;
+
+/**
+ * @author Silvio Giebl
```
[the remark this answers is not in the document]
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/util/FluentBuilder.java]: Thanks for the suggestion, will be added

[the remark this answers is not in the document]
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/util/FluentBuilder.java]: done

File: src/main/java/org/mqttbee/util/FluentBuilder.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public abstract class FluentBuilder<B, P> {
+
+    @NotNull
+    protected static <B, P> P done(@NotNull final B built, @Nullable final Function<B, P> parentConsumer) {
+        if (parentConsumer == null) {
+            throw new IllegalStateException("done must not be called on the root of a fluent builder");
```
[the remark this answers is not in the document]
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/util/FluentBuilder.java]: done

File: src/main/java/org/mqttbee/util/FluentBuilder.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.util.function.Function;
+
+/**
+ * @author Silvio Giebl
+ */
+public abstract class FluentBuilder<B, P> {
+
+    @NotNull
+    protected static <B, P> P done(@NotNull final B built, @Nullable final Function<B, P> parentConsumer) {
```
[the remark this answers is not in the document]
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/util/FluentBuilder.java]: As it simplifies the implementation of the `done` method in subclasses of builders

**Comment by christianhoff** [comment]
What about a fluent builder for Publish.qos?

File: src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (christianhoff) [reply src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java]: Probably we can reuse the inherited version. Have a look here: https://github.com/mqtt-bee/mqtt-bee/tree/improvement/fluent-builders-use-parent-consumer-from-base-class. It might need some improvement.

File: src/main/java/org/mqttbee/api/mqtt/MqttClientBuilder.java [hunk 6]
```diff
@@ -68,34 +69,49 @@ public MqttClientBuilder serverPort(final int port) {
     }
 
     @NotNull
-    public MqttClientBuilder useSsl() {
+    public MqttClientBuilder useSslDefaults() {
```
[the remark this answers is not in the document]
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/api/mqtt/MqttClientBuilder.java]: I renamed it because the name clashes with the new fluent builder method

File: src/main/java/org/mqttbee/api/mqtt/MqttClientBuilder.java [hunk 7]
```diff
@@ -68,34 +69,49 @@ public MqttClientBuilder serverPort(final int port) {
...
     }
 
     @NotNull
-    public MqttClientBuilder useWebSocket() {
+    public MqttClientSslConfigBuilder<? extends MqttClientBuilder> useSsl() {
+        return new MqttClientSslConfigBuilder<>(this::useSsl);
+    }
+
+    @NotNull
+    public MqttClientBuilder useWebSocketDefaults() {
```
[the remark this answers is not in the document]
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/api/mqtt/MqttClientBuilder.java]: I renamed it because the name clashes with the new fluent builder method

File: src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java [hunk 8]
```diff
@@ -18,35 +18,52 @@
...
+        return this;
     }
 
     @NotNull
     @Override
-    public MqttSharedTopicFilter multiLevelWildcard() {
-        return (MqttSharedTopicFilter) super.multiLevelWildcard();
+    public MqttSharedTopicFilter multiLevelWildcardAndBuild() {
+        super.multiLevelWildcardAndBuild();
+        return build();
```
Comment (christianhoff) [inline src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java]: build() is called two times. I think you have to call super.multiLevelWildcard() and make it protected.

File: src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java [hunk 9]
```diff
@@ -18,35 +18,52 @@
...
+    public MqttSharedTopicFilterBuilder<P> singleLevelWildcard() {
+        super.singleLevelWildcard();
+        return this;
     }
 
     @NotNull
     @Override
-    public MqttSharedTopicFilter multiLevelWildcard() {
-        return (MqttSharedTopicFilter) super.multiLevelWildcard();
+    public MqttSharedTopicFilter multiLevelWildcardAndBuild() {
```
Comment (christianhoff) [inline src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java]: Do we also need a multilevelWildcardAndDone() method?

File: src/main/java/org/mqttbee/util/FluentBuilder.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+     * Creates the builder's object and hands it over to its parent builder method.
+     * <p>
+     * This method must not be called on the root of a fluent builder. Consider calling {@link #build()} instead.
+     *
+     * @return the parent builder.
+     */
+    @NotNull
+    public P done() {
+        if (parentConsumer == null) {
+            throw new IllegalStateException("done must not be called on the root of a fluent builder");
```
Comment (christianhoff) [inline src/main/java/org/mqttbee/util/FluentBuilder.java]: provide a hint in the message of the exception as well, e.g. "Consinder calling build() instead.

File: src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java (hunk 8 above)
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java]: you are right, casting is maybe more clean
`return (MqttSharedTopicFilter) super.multiLevelWildcardAndBuild()`

File: src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java (hunk 9 above)
Comment (SgtSilvio) [reply src/main/java/org/mqttbee/api/mqtt/datatypes/MqttSharedTopicFilterBuilder.java]: No, `multilevelWildcardAndDone` is defined in MqttTopicFilterBuilder and we do not have to override it as it returns the parent builder.
